### DISCUSSION TITLE
rtl help

### DISCUSSION THREAD

**Discussion by ykrem** [typst/typst#4588] [Q&A]
long story short, I've been using typst for school for a few good months now. very satisfied with the language, yet due to me studying in Israel, I gotta use Hebrew (rtl). The thing is, neither the web app or vscode support proper rtl text editing, and it doesn't seem to change anytime soon. Any suggestions? Stuff like msword seems horrible for math compared to typst, and I'm willing to give up the visual aspect.

**Comment by Enivex** [comment]
@emilyyyylime you write Hebrew no?

**Reply by emilyyyylime** [reply]
I do and in fact I know @yonatankremer kek. Could've just DMed me o_O.

Anyways; in terms of an editor, there will be no good editor. I've come to accept it as a hard truth and honestly even if I think about how I might set out to specially design a programming-oriented text editor with bidirectional text support I don't know how I would do it. You would have to somehow have the editor display each nested scope (like an inline maths block etc.) surrounded by directional isolate characters (which is the primary solution offered by Unicode for controlling the behaviour of bidirectional text), and even that probably won't be enough, requiring some heuristic to decide whether a particular line should be laid out right-to-left or left-to-right.
Doing all that seamlessly while providing a developer-friendly editing experience in all other aspects sounds nigh impossible.

Unfortunately my only offer in that regard is to Just Cope. My current setup uses Helix inside COSMIC Terminal (for its bidirectional text support, though I've also heard good of Gnome-terminal and I've been meaning to try it out), tiled alongside a Zathura window for real-time rendering (with Typst-lsp, of course). Before finding COSMIC Terminal I used to use VSCode with the popular extensions and honestly it served me well enough. Some of the defaults aren't great but I'm sure you already know how to configure VSCode.

In terms of producing beautiful documents, Typst is hands down the best tool for the job. Unfortunately the latest release 0.11.0 comes with some critical bugs when it comes to inline maths within RTL text, but building from git or installing through cargo(?) and applying `#show math.equation: set text(dir: ltr)` to your documents should set you up for typesetting Hebrew with maths. I do hope this can return to being the default behaviour by the next release. (Well it technically never was the default but before equations being able to break the end result was the same).

My biggest tip is to use line breaks in your source code sparingly, they don't render in the compiled document, and having long lines wrap around under you can be a real mess when combined with bidirectional text. Hope this helps! (I should probably save this comment somewhere 🤔)

**Reply by Enivex** [reply]
I thought the RTL math bug was fixed in 0.11.1?

**Reply by Idobenhamo** [reply]
Can you please share a screenshot of rtl file inside of helix in the cosmic terminal?

**Reply by emilyyyylime** [reply]
@Idobenhamo here you go ^^ ![image.png](https://github.com/user-attachments/assets/1250b283-a694-4bb9-893d-f9428b810813)

@Enivex You're right, I was still using typst-lsp so I didn't have the latest features. After updating to tinymist it got fixed

**Reply by IgKh** [reply]
Hi @emilyyyylime I'm really interested in hearing more regarding your pessimism about the intractability of having a BiDi oriented programmer's editor. I've actually been working for quite some months now on exactly that idea. It feels to me that the most naive approach - give each line the base direction of its' first character with strong directionality (but allow to quickly toggle), inline math goes into a LRI isolate, inline variable references get a LRM to align the initial hashtag - go a long way. While it is a lot of control characters to type, keyboard macros for them make it convenient. I've been able to get through a year of an Economics major this way very comfortably, so I'm wondering if I missed some important aspect of it.

E.g your example (assuming I managed to read it correctly) looks like:
![Screenshot_20240825_233739](https://github.com/user-attachments/assets/58ac1fce-0e3d-4862-854e-ab40116bbce2)

While bolting programming features onto a rich text editor base was/is significant work, it seems much easier than trying to get a code editor control to do the correct text shaping and layout.

Thanks!

**Reply by ykrem** [reply]
holy shit it looks perfect😍
wml with infi 1 ;)

**Reply by emilyyyylime** [reply]
@IgKh it seems for your use you manually insert RTL embedding characters? I suppose that makes sense, but I was talking about a truly native experience, where you wouldn't need to enter any embeddings yourself (and they would not be present in the encoded file, either). Inserting them yourself is a great solution however, and it seems Katvan has a bright future :)

**Reply by IgKh** [reply]
Thanks @emilyyyylime, indeed the control characters need to inserted by the user although there are shortcuts in the application to try and smooth it  - i.e. `Ctrl+M` for inserting a pair of $ characters inside a LRI isolate (which is the 99.9999% case), and `Ctrl+Shift+I` for a quick menu with everything else.

Thinking about it, I think that a "truly native experience" isn't impossible. Qt's rich text engine, which implements the BiDi algorithm, has access to the format/style information as well as the text itself. However currently directionality is only a block (paragraph) level format option, and is only used to set the base directionality. I can conceive extending the character level format options to have a directionality attribute too, so that the BiDi algorithm implementation can consider a sequence of characters that have the same attribute as-if they were surrounded by the appropriate LRI/RLI <-> PDI pair. The editor can then set them based on the parsed syntax. But I feel like this is very niche and such a patch won't be accepted upstream to Qt.

**Reply by IgKh** [reply]
A brief update, for anyone interested. I know this is no longer the forum used, but commenting here anyway to keep things related.

I've been looking for some time for a way to implement what @emilyyyylime called "truly native experience" into Katvan. As kinda  expected, the discussion on extending Qt with the requisite capabilities directly got a fairly disinterested reception on the mailing list. I was looking for an alternative approach, using a parser-generated shadow layout for each line that is different from the actual edited text; this means there is no need to insert control characters and they don't appear in the file on-disk nor are they visible to the Typst compiler. 

There are still some small but annoying rendering glitches, but they don't seem to be workflow breaking and hopefully can be fixed as well. Otherwise the experience seems quite nice to me, and this should be in the next version of Katvan (or right now in master, for anyone who wants to try it out).

**Comment by rwmpelstilzchen** [comment]
היי 👋

There’s [Katvan](https://github.com/IgKh/katvan) by @IgKh, which is a text editor made with special care for the issue of intertwined RTL and LTR segments in Typst.

If you like its bare-bones approach, there you go. If like me you prefer to have a fully-fledged text editor at your fingertips, with all the modern goodness, give NeoVim a try. It plays extremely well with LSP ([tinymist](https://github.com/Myriad-Dreamin/tinymist), and supposedly also [typst-lsp](https://github.com/nvarner/typst-lsp) which I didn’t try); this upgrades the coding/writing experience remarkably. Anyway, regarding the BiDi matter in question, NeoVim’s approach inherits from Vim’s one, which is like so: either a window¹ is fully LTR, presenting each letter in the order it is encoded in the buffer from left to right, or fully RTL, from right to left. You can map a key combination or sequence to toggle between LTR and RTL modes². It takes some time to change your habits from what you are accustomed to in WYSIWYG editors, browser textboxes and almost everywhere nowadays to a this way of working, but IMHO it is worthwhile in the long run if you are going to type a lot in Typst (or LaTeX for that matter), intertwining RTL and LTR sequences. This way you see directly what appears in the file, unclouded by BiDi algorithms. If needed, you can split a buffer between two windows and set each the other directionality. I wrote my Master’s thesis with this approach, using LaTeX ([PDF, source files and meta information](https://old.digitalwords.net/mathesis/)), so I can testify for its efficiency with complex BiDi files (a Hebrew linguistic text involving a lot of Old English words and LaTeX commands in my case, not Hebrew + Typst’s math, but the principle is the same…). For completion, ages ago I wrote [this page](https://old.digitalwords.net/34/); most of what I wrote there is greatly outdated, but the principle of what I wrote about Hebrew there stays true.

¹ In the [*Vim sense](https://neovim.io/doc/user/windows.html).

² I use double space, as I (almost) never type two spaces sequentially:

```lua
vim.keymap.del("n", "<Space><Space>")
vim.keymap.set({ "n" }, "<Space><Space>", function()
  vim.opt.rl = not vim.opt.rl:get()
end, { desc = "Toggle directionality" })
```

**Reply by rwmpelstilzchen** [reply]
And as a side note… 😉

> I gotta use Hebrew (rtl)

https://media.giphy.com/media/F8SnqlpllPhss/giphy.gif

**Reply by rwmpelstilzchen** [reply]
And this is how the first sentence looks when you show the same buffer in two windows side by side, each with its own directionality.

![NeoVim-BiDi](https://github.com/user-attachments/assets/b7f3c4e2-cfff-4b4b-91e0-aac26017527a)

(⚠️ didn’t double check my typing; added [Hebrew maqafs](https://en.wikipedia.org/wiki/Hebrew_punctuation#Hyphen_and_maqaf))

**Reply by ykrem** [reply]
I can't seem to compile katvan on my mac, the "libarchive" dependency caused some trouble, and I don't know anything about cmake besides the error messages' guidance. Neovim isn't worth the switch from vscode for me atm.

**Reply by rwmpelstilzchen** [reply]
Can’t help you there, sorry. The last time I had a computer with macOS was before it was called ‘macOS’ (or ‘OS X’)… 😛

If you can’t use VS Code and Katvan, and won’t use NeoVim, perhaps Emacs would be a solution. I played a little with modern distros ([Doom Emacs](https://github.com/doomemacs/doomemacs) and [Spacemacs](https://www.spacemacs.org/)) and they seem impressive. Also here there is a fairly steep learning curve, so I don’t know whether you want to switch to Emacs either, atm.

Good luck, anyway!

**Reply by ykrem** [reply]
Thanks! Got katvan to work, was indeed a dependency path issue
